core/unit_test/hip/TestHIP_Memory_Requirements.cpp

```diff
@@ -0,0 +1,84 @@
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <Kokkos_Core.hpp>
+#include <TestHIP_Category.hpp>
+
+namespace Test {
+
+template <class HIPMemoryContainer>
+bool checkMemoryCoarseGrainedness(HIPMemoryContainer& container) {
+  auto size           = container.size();
+  auto allocationSize = HIPMemoryContainer::required_allocation_size(size);
+  hipMemRangeCoherencyMode memInfo;
+
+  KOKKOS_IMPL_HIP_SAFE_CALL(hipMemRangeGetAttribute(
+      &memInfo, sizeof(hipMemRangeCoherencyMode),
+      hipMemRangeAttributeCoherencyMode, container.data(), allocationSize));
+
+  return (hipMemRangeCoherencyModeCoarseGrain == memInfo);
+}
+
+#define KOKKOS_TEST_MEMORY_COARSEGRAINEDNESS(MEMORY_SPACE, DATATYPE, SIZE) \
+  {                                                                        \
+    Kokkos::View<DATATYPE*, MEMORY_SPACE> view(#MEMORY_SPACE, SIZE);       \
+    ASSERT_TRUE(view.is_allocated());                                      \
+    ASSERT_TRUE(checkMemoryCoarseGrainedness(view));                       \
+  }
+
+TEST(hip, memory_requirements) {
+  Kokkos::initialize();
+
+  KOKKOS_TEST_MEMORY_COARSEGRAINEDNESS(Kokkos::Experimental::HIPSpace, int, 10);
+  // KOKKOS_TEST_MEMORY_COARSEGRAINEDNESS(Kokkos::Experimental::HIPHostPinnedSpace,int,10);
```

Comment:
What's your intention here? Should we check this, too? Does it currently work?
I would prefer not to introduce commented code.

Reply:
If we want to make all out user-facing memory coarse-grained, we should test for it as the documentation is not the most reliable thing concerning this. But we could do this for the device memory in another pr, just as for HostPinned (although for HostPinned we need changes at the allocation, for the device memory it already works)

Reply:
I tend to agree with Daniel.  Put yourself in the shoes of the next person (possibly future you) that will go over these lines.  He will certainly appreciate a comment saying why the line is currently commented out and what the plan might be.